integration/plugin/mountpoint/mountpoint_unix_test.go

```diff
@@ -0,0 +1,899 @@
+						Exists: []mountpoint.StringMapKeyValuePattern{{
+							Key:   mountpoint.StringPattern{Exactly: "o"},
+							Value: mountpoint.StringPattern{Contains: "bind"},
+						}},
+					}},
+				},
+			},
+		},
+	}
+	return
+}
+
+func mountPointControllerForAnonymousBindMounts() (ctrl *testMountPointController) {
+	typeVolume := mountpoint.TypeVolume
+	ctrl = emptyMountPointController()
+
+	// matches -v /container
+	ctrl.propertiesRes = mountpoint.PropertiesResponse{
+		Success: true,
+		Patterns: []mountpoint.Pattern{
+			{
+				Type: &typeVolume,
+				Volume: mountpoint.VolumePattern{
+					Driver: []mountpoint.StringPattern{{Exactly: "local"}},
+					Options: []mountpoint.StringMapPattern{{
+						Not: true,
+						Exists: []mountpoint.StringMapKeyValuePattern{
+							{Key: mountpoint.StringPattern{Exactly: "o"}},
+							{Key: mountpoint.StringPattern{Exactly: "device"}},
+							{Key: mountpoint.StringPattern{Exactly: "type"}},
+						},
+					}},
+				},
+			},
+		},
+	}
+	return
+}
+
+func mountPointControllerForAllBindMounts() (ctrl *testMountPointController) {
+	typeBind := mountpoint.TypeBind
+	typeVolume := mountpoint.TypeVolume
+	ctrl = emptyMountPointController()
+
+	// matches all bind mounts
+	ctrl.propertiesRes = mountpoint.PropertiesResponse{
+		Success: true,
+		Patterns: []mountpoint.Pattern{
+			{Type: &typeBind},
+			{
+				Type: &typeVolume,
+				Volume: mountpoint.VolumePattern{
+					Driver: []mountpoint.StringPattern{{Exactly: "local"}},
+					Options: []mountpoint.StringMapPattern{{
+						Exists: []mountpoint.StringMapKeyValuePattern{{
+							Key:   mountpoint.StringPattern{Exactly: "o"},
+							Value: mountpoint.StringPattern{Contains: "bind"},
+						}},
+					}},
+				},
+			},
+		},
+	}
+	return
+}
+
+func mountPointControllerForMountsIntoContainersWithoutAllowLabel() (ctrl *testMountPointController) {
+	ctrl = emptyMountPointController()
+
+	// matches all mounts into containers without a label 'allow'
+	ctrl.propertiesRes = mountpoint.PropertiesResponse{
+		Success: true,
+		Patterns: []mountpoint.Pattern{
+			{Container: mountpoint.ContainerPattern{
+				Labels: []mountpoint.StringMapPattern{{
+					Not: true,
+					Exists: []mountpoint.StringMapKeyValuePattern{{
+						Key: mountpoint.StringPattern{Exactly: "allow"},
+					}},
+				}},
+			}},
+		},
+	}
+	return
+}
+
+func setupTest(c *testing.T) func() {
+	environment.ProtectImages(c, testEnv)
+
+	d = daemon.New(c, "", dockerdBinary, daemon.Config{
+		Experimental: testEnv.DaemonInfo.ExperimentalBuild,
+	})
+
+	mountPointController[0] = mountPointControllerForVBindMountOnly()
+	mountPointController[1] = mountPointControllerForVBindMountAndAllLocalVolumes()
+	mountPointController[2] = mountPointControllerForLocalVolumeBindMounts()
+	mountPointController[3] = mountPointControllerForAnonymousBindMounts()
+	mountPointController[4] = mountPointControllerForAllBindMounts()
+	mountPointController[5] = mountPointControllerForMountsIntoContainersWithoutAllowLabel()
```

Comment:
It seems that there are 6 controllers being setup, but some tests also modify the controllers. I believe not every test uses every controller, is that right?

I think this is going to be difficult to maintain. Instead of having everything setup for every test could you have a function like:

```go
func setupPlugin(controllers ...testMountPointController) { ... }
```

and call that on each test with just the controllers that it needs? That should also remove the need to modify the static controllers from individual tests, and will hopefully remove the indexing using numbers as well.

`setupPlugin()` can still add them to a static slice, and teardown can iterate over the slice to perform cleanup.